falcon/media/multipart.py

```diff
@@ -0,0 +1,387 @@
+                'unsupported charset: {}'.format(charset))
+        return self.data.decode(charset)
+
+    @property
+    def content_type(self):
+        # NOTE(vytas): RFC 7578, section 4.4.
+        #   Each part MAY have an (optional) "Content-Type" header field, which
+        #   defaults to "text/plain".
+        value = self._headers.get(b'content-type') or b'text/plain'
+        return value.decode('ascii')
+
+    @property
+    def filename(self):
+        if self._filename is None:
+
+            if self._content_disposition is None:
+                value = self._headers.get(b'content-disposition', b'')
+                self._content_disposition = cgi.parse_header(value.decode())
+
+            _, params = self._content_disposition
+
+            # NOTE(vytas): Supporting filename* as per RFC5987, as that has
+            #   been spotted in the wild, even though RFC 7578 forbids it.
+            match = _FILENAME_STAR_RFC5987.match(params.get('filename*', ''))
+            if match:
+                charset, value = match.groups()
+                charset = charset.lower()
+                if charset not in self._parse_options.supported_charsets:
+                    raise MultipartParseError(
+                        'unsupported charset: {}'.format(charset))
+                self._filename = unquote_to_bytes(value).decode(charset)
+            else:
+                value = params.get('filename')
+                if value is None:
+                    return None
+                self._filename = value
+
+        return self._filename
+
+    @property
+    def secure_filename(self):
+        return misc.secure_filename(self.filename)
+
+    @property
+    def name(self):
+        if self._name is None:
+
+            if self._content_disposition is None:
+                value = self._headers.get(b'content-disposition', b'')
+                self._content_disposition = cgi.parse_header(value.decode())
+
+            _, params = self._content_disposition
+            self._name = params.get('name')
+
+        return self._name
+
+    @property
+    def media(self):
+        if self._media is None:
+            handler = self._parse_options.media_handlers.find_by_media_type(
+                self.content_type, 'text/plain')
+            self._media = handler.deserialize(
+                self.stream, self.content_type, None)
+
+        return self._media
+
+
+class MultipartForm:
+
+    def __init__(self, stream, boundary, content_length, parse_options):
+        # if not isinstance(stream, BufferedStream):
+        if not hasattr(stream, 'read_until'):
+            if isinstance(stream, request_helpers.BoundedStream):
+                stream = BufferedStream(stream.stream.read, content_length)
+            else:
+                stream = BufferedStream(stream.read, content_length)
+
+        self._stream = stream
+        self._boundary = boundary
+        self._dash_boundary = b'--' + boundary
```

Comment:
We might want to add a reference to the [RFC 7578#4.1](https://tools.ietf.org/html/rfc7578#section-4.1) for why this is necessary.

Reply:
Thanks, will do.